### PR TITLE
Switch windows workflow to run on Azure hosted ARC runner (#18859)

### PR DESCRIPTION
**Pull request by Eliasj42**
Progress on https://github.com/iree-org/iree/issues/18813

workflow running: https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374

---------

### CODE REVIEW

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: All generic (non-project-specific) setup should be moved to the Dockerfile. I'm less concerned about steps taking up to 1 minute, but the 15 minute MSVC install will need to be moved to ahead of time. Our target for presubmit should be 10-20 minutes per run, including the source build (warm cache). We can budget for 5 minutes or so of overhead (git checkout, installing deps, etc.).

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:25]: How many cores is this configured with right now? https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374 is making slow progress.

@saienduri said that we might only have quota for small runners, but we need to make a case for larger runners based on utilization or something? If so, that's silly - we know that we need 32 (or even 64/96) core runners and we already have data to back that up. We _could_ switch the current nightly build to these new runners if it is necessary to make a case for quota increases, but we're currently paying $0 for free runners that take 4h30m. Paying for runners that are similarly slow doesn't make sense to me :P

cc @amd-chrissosa

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml]: (All of these steps should move to the Dockerfile too. Some amount of workflow scripting is okay if it unblocks these new runners faster... _if the overhead is low enough_)

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:61]: Is this necessary? The GitHub runner should do some automatic cleanup already. We can tolerate some scripting like this if it unblocks other work, but this does increase workflow complexity. This also fails to clean up the directory if the workflow fails or is cancelled (it would need a `if(always)`... but even that can fail to run if the runner goes offline before running that step).

Maybe keep the build directory in the github workspace? Why did you put things under `C:\mnt\azure\`?

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (ScottTodd) [inline .github/workflows/ci_windows_x64_msvc.yml:57]: Build error: https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374#step:13:8554

```
[7708/8119] Linking CXX shared library tools\IREECompiler.dll
FAILED: tools/IREECompiler.dll lib/IREECompiler.lib 
C:\Windows\system32\cmd.exe /C "cd . && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_dll --intdir=compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir --rc="C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\rc.exe" --mt="C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\mt.exe" --manifests  -- "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\bin\Hostx64\x64\link.exe" /nologo @CMakeFiles\iree_compiler_API_SharedImpl.rsp  /out:tools\IREECompiler.dll /implib:lib\IREECompiler.lib /pdb:tools\IREECompiler.pdb /dll /version:0.0 /machine:x64 -fuse-ld=lld /debug /INCREMENTAL  -natvis:C:/home/runner/_work/iree/iree/runtime/iree.natvis && cd ."
LINK Pass 1: command "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\iree_compiler_API_SharedImpl.rsp /out:tools\IREECompiler.dll /implib:lib\IREECompiler.lib /pdb:tools\IREECompiler.pdb /dll /version:0.0 /machine:x64 -fuse-ld=lld /debug /INCREMENTAL -natvis:C:/home/runner/_work/iree/iree/runtime/iree.natvis /MANIFEST /MANIFESTFILE:compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir/intermediate.manifest compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir/manifest.res" failed (exit code 1140) with the following output:
LINK : warning LNK4044: unrecognized option '/fuse-ld=lld'; ignored

LINK : warning LNK4044: unrecognized option '/fuse-ld=lld'; ignored

   Creating library lib\IREECompiler.lib and object lib\IREECompiler.exp

LINK : fatal error LNK1318: Unexpected PDB error; LIMIT (12) 'C:\mnt\azure\build-windows\tools\IREECompiler.pdb'
```

Hmmm... could be multiple reasons for that. First thing I'd suspect is out of disk / RAM.

**Review by ScottTodd — COMMENTED** [review]
BTW this is great progress, thanks so much! Getting the runners set up with the general set of software installed is a huge milestone.

**Comment by ScottTodd** [comment]
If iterating on a single workflow via a pull request, consider using `skip-ci` or the other git trailers to avoid running unrelated workflows: https://iree.dev/developers/general/contributing/#ci-behavior-manipulation . You can also use `workflow_dispatch` to run just when you want and avoid sending notifications for every test you run

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (saienduri) [reply .github/workflows/ci_windows_x64_msvc.yml:25]: Currently, this cluster is using instances with 8 cpu cores

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (saienduri) [reply .github/workflows/ci_windows_x64_msvc.yml:61]: For some reason, the default storage drive is hard set to less space than needed and flipping different switches/requesting more memory wasn't able to resolve it for the pod instances. So, we decided to use a storage mount that we have full control over

File: .github/workflows/ci_windows_x64_msvc.yml
Comment (Eliasj42) [reply .github/workflows/ci_windows_x64_msvc.yml:57]: Hi, sorry I've been neglecting this page while troubleshooting the above issue.  I fixed it, the only remaining issue is a certification error 

ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1007)

https://github.com/iree-org/iree/actions/runs/11469526800/job/31916920407

For some reason when I install msvc in the Docker container, the workflow can't find it.

**Comment by ScottTodd** [comment]
What steps do you have planned for this, @Eliasj42 ? These look like the latest logs: https://github.com/iree-org/iree/actions/runs/11522453338/job/32078360486 (from the full history here: https://github.com/iree-org/iree/actions/workflows/ci_windows_x64_msvc.yml)

Per https://github.com/iree-org/iree/issues/18813, we're doing this work to get the Windows workflow running on presubmit, in 20-30 minutes. Do you want to start with this incremental change to get the 5h nightly build on GitHub-hosted runners down to 3h on these self-hosted runners and start collecting data? To reach the 20-30 minute target, we'll need to (at a minimum) get larger runner machines in the cluster and move the slow setup steps (e.g. installing MSVC) from the workflow to the cluster configuration (via the dockerfile).

I'm not really comfortable landing with the "Clean up build dir" or "Installing MSVC requirements" steps when those should be part of the runner environment itself... any steps that are tightly coupled with the runner definitions are going to be difficult for any maintainers/contributors to understand and work with.

(btw, please click "re-request review" when you want another round of review comments)
image

**Comment by Eliasj42** [comment]
> What steps do you have planned for this, @Eliasj42 ? These look like the latest logs: https://github.com/iree-org/iree/actions/runs/11522453338/job/32078360486 (from the full history here: https://github.com/iree-org/iree/actions/workflows/ci_windows_x64_msvc.yml)
> 
> Per #18813, we're doing this work to get the Windows workflow running on presubmit, in 20-30 minutes. Do you want to start with this incremental change to get the 5h nightly build on GitHub-hosted runners down to 3h on these self-hosted runners and start collecting data? To reach the 20-30 minute target, we'll need to (at a minimum) get larger runner machines in the cluster and move the slow setup steps (e.g. installing MSVC) from the workflow to the cluster configuration (via the dockerfile).
> 
> I'm not really comfortable landing with the "Clean up build dir" or "Installing MSVC requirements" steps when those should be part of the runner environment itself... any steps that are tightly coupled with the runner definitions are going to be difficult for any maintainers/contributors to understand and work with.
> 
> (btw, please click "re-request review" when you want another round of review comments) image

For some reason the setting up MSVC causes an error when I run it in the docker file, I'm looking into that, but I don't know if this is a priority to get merged in the meantime.  Also, I can experiment more with the user pool to get more space if the mounted hardrive isn't a good solution

**Comment by ScottTodd** [comment]
> but I don't know if this is a priority to get merged in the meantime.

The end goal is 20-30 minute Windows builds running on every commit. If we merge this as-is, we'll go from a 5 hour build on free runners that GitHub manages to a 3 hour build on runners we pay for and manage. Having functional self-hosted runners is certainly a useful engineering milestone, but I'd need to see a plan for how we bridge that gap and get value out of that extra runner spend before I'd approve the tech debt (runner-specific steps in the workflow file). If we need to demonstrate usage to some internal team before they grant us quota (which seems completely unacceptable... there's no way that we can settle for 8 CPU core runners and we _already have data proving that_), then such experiments can live on a branch or in a fork.

**Pull request closed without merging**
